include/cell.tcc

```diff
@@ -979,37 +979,106 @@ bool mpm::Cell<Tdim>::assign_velocity_constraint(unsigned face_id, unsigned dir,
   return status;
 }
 
-//! Compute all face normals 2d
+//! Apply velocity constraints
+template <unsigned Tdim>
+void mpm::Cell<Tdim>::apply_velocity_constraints() {
+
+  // Assum phase = 0 - TODO: make it general
+  const unsigned phase = 0;
+
+  // Copy nodal velocities and accelerations
+  for (unsigned i = 0; i < this->nfunctions(); ++i) {
```

Comment:
Why do we need to copy this information from the nodes, is it not sufficient to tell the nodes what the velocities and accelerations should be? You shouldn't need to know what the velocity currently is, to do this.